### PR TITLE
[5.2] Error handling on extension update

### PR DESCRIPTION
**Pull request by BrainforgeUK**
Pull Request for Issue #43319  .

### Summary of Changes

Catch errors from the extension being updated and abort update with a sensible message.

### Testing Instructions

(a) Installed this Plugin.
[plg_updatetest1.zip](https://github.com/joomla/joomla-cms/files/15048374/plg_updatetest1.zip)

(b) **Enable the plugin.**

(c) Go to Joomla updater and check for updates.

(d) Then select the plugin installed above PLG_HIKASHOPSHIPPING_BFEXTRAFEATURES and update it.

### Actual result BEFORE applying this Pull Request

Error 0

### Expected result AFTER applying this Pull Request

Error 1

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by brianteeman** [comment]
Cannot be tested
image

**Comment by BrainforgeUK** [comment]
Try this test plugin - screenshots before applying the pull request changes.
[plg_updatetest2.zip](https://github.com/joomla/joomla-cms/files/15049149/plg_updatetest2.zip)

Error 10
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Error 15
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Error 12
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Error 16

**Comment by brianteeman** [comment]
image

**Comment by BrainforgeUK** [comment]
Test does not work on Unix system! - must be filename case, investigating.

**Comment by brianteeman** [comment]
I am using windows

**Comment by BrainforgeUK** [comment]
Sorted, no change to the above needed, can go ahead with testing.
Appologizes for inconvenience, update server issue, not looked at that for a long time!

**Comment by brianteeman** [comment]
With Debug Disabled
image

With Debug Enabled
image


PLEASE check your work before submitting it

**Comment by BrainforgeUK** [comment]
Changed.

Sorry, copied the changed code across and forgot the _use_ statement.
Failed to check the _files changed_ count - thought I had copied the language text across.

File: administrator/components/com_installer/src/Model/UpdateModel.php [hunk 1]
```diff
@@ -363,10 +364,22 @@ public function update($uids, $minimumStability = Updater::STABILITY_STABLE)
...
+                $res = $this->install($update);
+            }
+            catch (\Throwable $t)
+            {
+                $res = false;
+
+                Factory::getApplication()->enqueueMessage(Text::sprintf('COM_INSTALLER_UPDATE_ERROR',
+                    $instance->name,
+                    $t->getMessage(),
+                    (JDEBUG ? str_replace(JPATH_ROOT, 'JROOT', Path::clean($t->getFile())) . ':' .  $t->getLine() : '')),
```
Comment (Quy) [inline administrator/components/com_installer/src/Model/UpdateModel.php]:
```suggestion
                    (JDEBUG ? str_replace(JPATH_ROOT, 'JROOT', Path::clean($t->getFile())) . ':' . $t->getLine() : '')),
```

**Comment by brianteeman** [comment]
You also forgot to check the language string you used and the string you created.

File: administrator/language/en-GB/com_installer.ini [hunk 2]
```diff
@@ -263,6 +263,7 @@ COM_INSTALLER_UPDATE_FILTER_SEARCH_LABEL="Search Extensions with Updates"
 COM_INSTALLER_UPDATE_FORM_EDIT="Edit Update Site"
 COM_INSTALLER_UPDATE_MISSING_DOWNLOADKEY_LABEL_N="%d extension updates cannot be installed until you <a href=\"%s\">enter their corresponding Download Keys</a>."
 COM_INSTALLER_UPDATE_MISSING_DOWNLOADKEY_LABEL_N_1="One extension update cannot be installed until you <a href=\"%2$s\">enter its Download Key</a>."
+COM_INSTALLER_ERROR="Error updating: %1$s<br>%2$s<br>%3$s<br>"
```
Comment (richard67) [inline administrator/language/en-GB/com_installer.ini]: 1. Language strings should be alpha ordered by their key.
2. In the code you use `COM_INSTALLER_UPDATE_ERROR`, so I assume they key needs to be changed here to that key.

File: administrator/components/com_installer/src/Model/UpdateModel.php [hunk 3]
```diff
@@ -363,10 +364,22 @@ public function update($uids, $minimumStability = Updater::STABILITY_STABLE)
...
+            catch (\Throwable $t)
+            {
+                $res = false;
+
+                Factory::getApplication()->enqueueMessage(Text::sprintf('COM_INSTALLER_UPDATE_ERROR',
+                    $instance->name,
+                    $t->getMessage(),
+                    (JDEBUG ? str_replace(JPATH_ROOT, 'JROOT', Path::clean($t->getFile())) . ':' . $t->getLine() : '')),
+                    'error');
+            }
```
Comment (richard67) [inline administrator/components/com_installer/src/Model/UpdateModel.php]:
```suggestion
            } catch (\Throwable $t) {
                $res = false;

                Factory::getApplication()->enqueueMessage(
                    Text::sprintf(
                        'COM_INSTALLER_UPDATE_ERROR',
                        $instance->name,
                        $t->getMessage(),
                        (JDEBUG ? str_replace(JPATH_ROOT, 'JROOT', Path::clean($t->getFile())) . ':' . $t->getLine() : '')
                    ),
                    'error'
                );
            }
```
Fix PHPCS.

**Comment by BrainforgeUK** [comment]
Retested as: https://github.com/joomla/joomla-cms/pull/43321#issuecomment-2067737115

With changes made the message is now:

restest2

**Comment by BrainforgeUK** [comment]
Regardless of who is to blame for the error what I propose is better and more useful than what Joomla currently displays.

This is actually not a plugin bug - the plugin is used by and depends on a 3rd party extension, it is that dependency which causes the problem when the update is attempted with the plugin enabled. Could modify such plugins to protect against such errors but I see that as unnecessary code in the plugin when Joomla can handle it more gracefully.

I suppose could add another line to the error message:
_"Please contact the extension developer."_

Could enhance further by adding contact information from current XML?

**Comment by brianteeman** [comment]
The question is. Does the new message help the end user more than the original message. For me the answer is no

**Comment by BrainforgeUK** [comment]
The end user is an admin who hopefully knows the extensions installed on their site.

The current error is missleading as it implies it is a bug in the Joomla installer ... which led to me being asked a question which started me on this.

**Comment by chmst** [comment]
I think that this helps "Please contact the extension developer of the Plugin ...  ". 
Then I know that it is not a bug in Joomla or in my confirguration and don't waste time with debugging.

**Comment by brianteeman** [comment]
@chmst the _problem_ is that according to @BrainforgeUK the problem is not with the developer of this plugin but with the developer of a completely different extension upon which it depends - so which extension developer am I going to contact? For me that is not the problem. The problem is that the extension that is being installed is not checking for the presence of the extension it is dependent upon. ergo nothing sensible can be changed in core to fix an extension developers errors

**Comment by chmst** [comment]
@brianteeman - thanks, agreed.

**Comment by brianteeman** [comment]
there is only so much that Joomla can do. Extension developers need to take responsibility for their own code being installed correctly.

**Comment by fgsw** [comment]
should this PR get test or not?

**Comment by BrainforgeUK** [comment]
I do't see why not.
Improves on what is there already but as indicated in previous comments it is not perfect. There can never be a perfect solution.

**Comment by fgsw** [comment]
> I do't see why not.

I don't see positive comments. Don't wanna waste my time.

**Comment by exlemor** [comment]
I have tested this item :red_circle: unsuccessfully on bb1c6aa4acc1d4944f815b6f0950579616ac4582<br /><br />Hi BrainForgeUK, I have really tried to test this during PBF 2024, and I am sorry, but neither pig_updatetest1.zip nor png_updatetest2.zip yield the results that you show/want... perhaps because we are testing it now against Joomla 5.2.0-beta2-dev but couldn't get it to work out ;(<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43321">issues.joomla.org/tracker/joomla-cms/43321</a>.</sub>

**Comment by TLWebdesign** [comment]
I have tested this item :white_check_mark: successfully on bb1c6aa4acc1d4944f815b6f0950579616ac4582<br /><br />Tested successfully. Seems like a no brainer to me to improve this.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43321">issues.joomla.org/tracker/joomla-cms/43321</a>.</sub>

**Comment by pe7er** [comment]
I have tested this item :white_check_mark: successfully on bb1c6aa4acc1d4944f815b6f0950579616ac4582<br /><br />Works as described.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43321">issues.joomla.org/tracker/joomla-cms/43321</a>.</sub>

**Comment by TLWebdesign** [comment]
On Joomla 5.1.3 i'm not seeing the path. I tested it both on a localhost and on a live test server. 
See attached screenshot:
Scherm­afbeelding 2024-08-24 om 12 45 57

**Comment by brianteeman** [comment]
@TLWebdesign as this PR is for 5.1 and not 5.2 then you should be marking this as an unsuccessful test

**Comment by TLWebdesign** [comment]
> @TLWebdesign as this PR is for 5.1 and not 5.2 then you should be marking this as an unsuccessful test

should i remove the comment with the successful test to get rid of it or how do i retract my successful test?

**Comment by TLWebdesign** [comment]
I have tested this item :red_circle: unsuccessfully on bb1c6aa4acc1d4944f815b6f0950579616ac4582<br /><br />On 5.1.3 i'm not getting the path to the file to show. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43321">issues.joomla.org/tracker/joomla-cms/43321</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by TLWebdesign** [comment]
I have tested this item :white_check_mark: successfully on bb1c6aa4acc1d4944f815b6f0950579616ac4582<br /><br />Tested Succesfully on 5.2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43321">issues.joomla.org/tracker/joomla-cms/43321</a>.</sub>

File: administrator/language/en-GB/com_installer.ini [hunk 4]
```diff
@@ -263,6 +263,7 @@ COM_INSTALLER_UPDATE_FILTER_SEARCH_LABEL="Search Extensions with Updates"
 COM_INSTALLER_UPDATE_FORM_EDIT="Edit Update Site"
 COM_INSTALLER_UPDATE_MISSING_DOWNLOADKEY_LABEL_N="%d extension updates cannot be installed until you <a href=\"%s\">enter their corresponding Download Keys</a>."
 COM_INSTALLER_UPDATE_MISSING_DOWNLOADKEY_LABEL_N_1="One extension update cannot be installed until you <a href=\"%2$s\">enter its Download Key</a>."
+COM_INSTALLER_UPDATE_ERROR="Error updating: %1$s<br>%2$s<br>%3$s<br>"
```
Comment (Quy) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_UPDATE_ERROR="Error updating: %1$s<br>%2$s<br>%3$s<br>"
COM_INSTALLER_UPDATE_FORM_EDIT="Edit Update Site"
COM_INSTALLER_UPDATE_MISSING_DOWNLOADKEY_LABEL_N="%d extension updates cannot be installed until you <a href=\"%s\">enter their corresponding Download Keys</a>."
COM_INSTALLER_UPDATE_MISSING_DOWNLOADKEY_LABEL_N_1="One extension update cannot be installed until you <a href=\"%2$s\">enter its Download Key</a>."
```

File: administrator/language/en-GB/com_installer.ini [hunk 5]
```diff
@@ -260,6 +260,7 @@ COM_INSTALLER_UNINSTALL_SUCCESS="Uninstalling the %s was successful."
 COM_INSTALLER_UNPACK_ERROR="Failed to extract file: %s"
 COM_INSTALLER_UPDATE_FILTER_SEARCH_DESC="Search in extension name. Prefix with ID:, UID: or EID: to search for an update ID, update site ID or extension ID."
 COM_INSTALLER_UPDATE_FILTER_SEARCH_LABEL="Search Extensions with Updates"
+COM_INSTALLER_UPDATE_ERROR="Error updating: %1$s<br>%2$s<br>%3$s<br>"
```
Comment (Quy) [inline administrator/language/en-GB/com_installer.ini]:
```suggestion
COM_INSTALLER_UNPACK_ERROR="Failed to extract file: %s"
COM_INSTALLER_UPDATE_ERROR="Error updating: %1$s<br>%2$s<br>%3$s<br>"
COM_INSTALLER_UPDATE_FILTER_SEARCH_DESC="Search in extension name. Prefix with ID:, UID: or EID: to search for an update ID, update site ID or extension ID."
COM_INSTALLER_UPDATE_FILTER_SEARCH_LABEL="Search Extensions with Updates"
```

File: administrator/components/com_installer/src/Model/UpdateModel.php [hunk 6]
```diff
@@ -358,10 +359,22 @@ public function update($uids, $minimumStability = Updater::STABILITY_STABLE)
...
+            try {
+                $this->preparePreUpdate($update, $instance);
 
-            // Install sets state and enqueues messages
-            $res = $this->install($update);
+                // Install sets state and enqueues messages
+                $res = $this->install($update);
+            }
+            catch (\Throwable $t)
+            {
```
Comment (Quy) [inline administrator/components/com_installer/src/Model/UpdateModel.php]:
```suggestion
            } catch (\Throwable $t) {
```

File: administrator/components/com_installer/src/Model/UpdateModel.php [hunk 7]
```diff
@@ -358,10 +359,20 @@ public function update($uids, $minimumStability = Updater::STABILITY_STABLE)
...
+                // Install sets state and enqueues messages
+                $res = $this->install($update);
+            } catch (\Throwable $t) {
+                $res = false;
+
+                Factory::getApplication()->enqueueMessage(Text::sprintf('COM_INSTALLER_UPDATE_ERROR',
+                    $instance->name,
+                    $t->getMessage(),
+                    (JDEBUG ? str_replace(JPATH_ROOT, 'JROOT', Path::clean($t->getFile())) . ':' . $t->getLine() : '')),
+                    'error');
```
Comment (Quy) [inline administrator/components/com_installer/src/Model/UpdateModel.php]:
```suggestion
                Factory::getApplication()->enqueueMessage(
                    Text::sprintf(
                        'COM_INSTALLER_UPDATE_ERROR',
                        $instance->name,
                        $t->getMessage(),
                        (JDEBUG ? str_replace(JPATH_ROOT, 'JROOT', Path::clean($t->getFile())) . ':' . $t->getLine() : '')
                    ),
                    'error'
                );
```

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43321">issues.joomla.org/tracker/joomla-cms/43321</a>.</sub>

**Comment by pe7er** [comment]
Thanks @BrainforgeUK !

**Pull request merged**